compiler/modules/types/list.bal

```diff
@@ -572,11 +560,56 @@ function listSubtypeIntersect(Context cx, SubtypeData t1, SubtypeData t2) return
 }
 
 function listSubtypeDiff(Context cx, SubtypeData t1, SubtypeData t2) returns SubtypeData {
-    Bdd b = memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
-    if b != false && cx.listMemo[b] == () {
-        //io:println("fresh BDD: ", bddToString(b));
+    MemoBddCache cache = new (cx.listMemo);
+    // This is checking if t1 and t2 are disjoint (i.e intersection(t1, t2) is empty). If so diff(t1, t2) == t1.
+    if listIsEmptySimple(cx, bddIntersect(cache, <Bdd> t1, <Bdd> t2)) is true {
+        return t1;
+    }
+    return memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
+}
+
+function listIsEmptySimple(Context cx, Bdd bdd) returns boolean? {
+    BddMemo? m = cx.listMemo[bdd];
+    if m !is () && m.empty is boolean {
+        return <boolean>m.empty;
+    }
+    boolean empty = bddEvery(cx, bdd, (), (), listFormulaIsDefinitelyEmpty);
+    if empty == true {
+        cx.listMemo.put({ bdd, empty }); // handle both memoized values being () and not in memo
+        return true;
+    }
+    return ();
+}
+
+function listFormulaIsDefinitelyEmpty(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
+    if neg !is () || pos is () {
```

Comment:
As a matter of style, please use `==` instead of `is` for comparing against constants.

Reply:
fixed